### PR TITLE
Use ByteBufferStreamInput to Stream Byte Arrays

### PR DESCRIPTION
**Pull request by original-brownbear**
For bulk operations that fall back to hotspot intrinsic code (reading short, int, long)
using this stream brings a massive speedup. The added benchmark for reading `long` values
sees a ~100x speedup in local benchmarking and the vLong read benchmark still sees a slightly
under ~10x speedup.
Also, this PR moves creation of the `StreamInput` out of the hot benchmark loop for all the bytes
reference benchmarks to make the benchmark less noisy and more practically useful.
(the `readLong` case using intrinsic operations is so fast that it wouldn't even show up in a profile relative to
instantiating the stream otherwise).

Relates work in #71181 (I think given how fast the intrinsic ops are there's also significant potential for further speedups by using them in the multi-page `BytesReferenceStreamInput` in a follow-up)

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/io/stream/ByteBufferStreamInput.java [hunk 1]
```diff
@@ -49,31 +50,30 @@ public int read(byte[] b, int off, int len) throws IOException {
...
 
     @Override
     public short readShort() throws IOException {
         try {
             return buffer.getShort();
         } catch (BufferUnderflowException ex) {
-            EOFException eofException = new EOFException();
-            eofException.initCause(ex);
-            throw eofException;
+            throw throwEOFException(ex);
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/common/io/stream/ByteBufferStreamInput.java:76]: `throw throw`? Maybe `throw newEofException`?

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 2]
```diff
@@ -1319,7 +1319,7 @@ public static StreamInput wrap(byte[] bytes) {
     }
 
     public static StreamInput wrap(byte[] bytes, int offset, int length) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:1321]: Wow we call this method a lot!

I'm curious to see what this change will translate to in terms of raw performance in general use.

**Review by nik9000 — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/io/stream/ByteBufferStreamInput.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/io/stream/ByteBufferStreamInput.java:76]: ++

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:1321]: I think this is a big case of "it depends". The good news is probably that all operations will see a speedup (I ad hoc benchmarked reading strings etc. and there's some speedup there). On the other hand you will not see the single array stream over very large arrays too often (if you did in some spot, allocating those huge arrays should probably be fixed). But I think on balance even if it doesn't speed up too many operations end-to-end, it certainly does improve the load on the transport thread for some medium size messages which is always nice for stability :)

**Comment by original-brownbear** [comment]
Thanks Nik!

**Pull request merged**